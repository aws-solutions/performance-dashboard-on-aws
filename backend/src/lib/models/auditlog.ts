/*
 *  Copyright Amazon.com, Inc. or its affiliates. All Rights Reserved.
 *  SPDX-License-Identifier: Apache-2.0
 */

export enum ItemEvent {
    Create = "Create",
    Update = "Update",
    Delete = "Delete",
}

export type ModifiedProperty = {
    property: string;
    oldValue: any;
    newValue: any;
};

export interface AuditLog {
    itemId: string;
    timestamp: Date;
    userId: string;
    event: ItemEvent;
    modifiedProperties?: ModifiedProperty[];
}

// Represents a dynamodb item in the Audit Trail table
export interface AuditLogItem {
    pk: string;
    sk: string;
    type: string;
    userId: string;
    event: ItemEvent;
    modifiedProperties?: ModifiedProperty[];
}

export interface DashboardAuditLog extends AuditLog {
    version: number;
}

// Represents a dynamodb item in the Audit Trail table that
// contains attributes specific to Dashboard audit logs.
export interface DashboardAuditLogItem extends AuditLogItem {
    version: number;
}
