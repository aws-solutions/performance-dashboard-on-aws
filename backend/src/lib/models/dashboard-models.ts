export type Dashboard = {
    id: string,
    name: string,
    topicAreaId: string,
    topicAreaName: string,
    description: string,
    createdBy: string,
};

export type DashboardList = Array<Dashboard>;

export interface DashboardItem {
    pk: string,
    sk: string,
    type: string,
    dashboardName: string,
    topicAreaId: string,
    topicAreaName: string,
    description: string,
    createdBy: string,
};