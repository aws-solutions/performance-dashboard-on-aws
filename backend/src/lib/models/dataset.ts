/*
 *  Copyright Amazon.com, Inc. or its affiliates. All Rights Reserved.
 *  SPDX-License-Identifier: Apache-2.0
 */

export enum SourceType {
    IngestApi = "IngestApi",
    FileUpload = "FileUpload",
}

export enum DatasetSchema {
    None = "None",
    Metrics = "Metrics",
}

export interface Dataset {
    id: string;
    createdBy: string;
    fileName: string;
    s3Key: {
        raw: string;
        json: string;
    };
    updatedAt: Date;
    sourceType: SourceType;
    schema?: DatasetSchema;
}

export type DatasetContent = object[];
export type DatasetList = Dataset[];

export interface DatasetItem {
    pk: string;
    sk: string;
    type: string;
    createdBy: string;
    fileName: string;
    s3Key: {
        raw: string;
        json: string;
    };
    updatedAt: string;
    sourceType: string;
    schema?: string;
}
