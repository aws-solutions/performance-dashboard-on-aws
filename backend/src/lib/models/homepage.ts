/*
 *  Copyright Amazon.com, Inc. or its affiliates. All Rights Reserved.
 *  SPDX-License-Identifier: Apache-2.0
 */

export interface Homepage {
    title: string;
    description: string;
    updatedAt: Date;
}

export interface HomepageItem {
    pk: string;
    sk: string;
    type: string;
    updatedAt?: string;
    title: string;
    description: string;
}
