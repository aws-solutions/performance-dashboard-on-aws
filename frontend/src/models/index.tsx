/*
 *  Copyright Amazon.com, Inc. or its affiliates. All Rights Reserved.
 *  SPDX-License-Identifier: Apache-2.0
 */

export type TopicArea = {
    id: string;
    name: string;
    createdBy: string;
    dashboardCount: number;
};

export const TopicAreaSortingCriteria = (a: TopicArea, b: TopicArea) => (a.name > b.name ? 1 : -1);

export type PublicTopicArea = {
    id: string;
    name: string;
    dashboards?: Array<Dashboard | PublicDashboard>;
};

export enum DashboardState {
    Draft = "Draft",
    Published = "Published",
    Archived = "Archived",
    PublishPending = "PublishPending",
    Inactive = "Inactive",
}

export enum SourceType {
    IngestApi = "IngestApi",
    FileUpload = "FileUpload",
}

export enum DatasetType {
    DynamicDataset = "DynamicDataset",
    StaticDataset = "StaticDataset",
    CsvFileUpload = "CsvFileUpload",
    CreateNew = "CreateNew",
}

export type Dashboard = {
    id: string;
    name: string;
    version: number;
    parentDashboardId: string;
    topicAreaId: string;
    topicAreaName: string;
    displayTableOfContents: boolean;
    /**
     * @type {any}
     * @deprecated no longer supported
     */
    tableOfContents?: any;
    description?: string;
    releaseNotes?: string;
    widgets: Array<Widget>;
    state: string;
    updatedAt: Date;
    createdBy: string;
    submittedBy?: string;
    publishedBy?: string;
    archivedBy?: string;
    friendlyURL?: string;
    queryMatches?: Array<string>;
};

export type PublicDashboard = {
    id: string;
    name: string;
    topicAreaId: string;
    topicAreaName: string;
    displayTableOfContents: boolean;
    /**
     * @type {any}
     * @deprecated no longer supported
     */
    tableOfContents?: any;
    description?: string;
    widgets: Array<Widget>;
    updatedAt: Date;
    friendlyURL?: string;
    queryMatches?: Array<string>;
};

export type DashboardVersion = {
    id: string;
    version: number;
    state: DashboardState;
    friendlyURL?: string;
};

export type DashboardAuditLog = {
    itemId: string;
    timestamp: Date;
    version: number;
    userId: string;
    event: "Create" | "Update" | "Delete";
    modifiedProperties?: Array<{
        property: string;
        oldValue: any;
        newValue: any;
    }>;
};

export enum WidgetType {
    Text = "Text",
    Chart = "Chart",
    Table = "Table",
    Image = "Image",
    Metrics = "Metrics",
    Section = "Section",
}

export enum ChartType {
    LineChart = "LineChart",
    ColumnChart = "ColumnChart",
    BarChart = "BarChart",
    PartWholeChart = "PartWholeChart",
    PieChart = "PieChart",
    DonutChart = "DonutChart",
}

export interface Widget {
    id: string;
    name: string;
    widgetType: string;
    order: number;
    updatedAt: Date;
    dashboardId: string;
    content: any;
    showTitle: boolean;
    section?: string;
}

export interface ChartWidget extends Widget {
    content: {
        title: string;
        chartType: ChartType;
        datasetId: string;
        summary: string;
        summaryBelow: boolean;
        datasetType?: DatasetType;
        s3Key: {
            raw: string;
            json: string;
        };
        columnsMetadata: ColumnMetadata[];
        sortByColumn?: string;
        sortByDesc?: boolean;
        horizontalScroll?: boolean;
        stackedChart?: boolean;
        significantDigitLabels: boolean;
        dataLabels: boolean;
        computePercentages: boolean;
        showTotal: boolean;
    };
}

export interface TableWidget extends Widget {
    content: {
        title: string;
        datasetId: string;
        summary: string;
        summaryBelow: boolean;
        datasetType?: DatasetType;
        s3Key: {
            raw: string;
            json: string;
        };
        columnsMetadata: ColumnMetadata[];
        sortByColumn?: string;
        sortByDesc?: boolean;
        significantDigitLabels: boolean;
        displayWithPages: boolean;
    };
}

export interface ImageWidget extends Widget {
    content: {
        title: string;
        imageAltText: string;
        summary?: string;
        summaryBelow: boolean;
        s3Key: {
            raw: string;
        };
        filename: string;
        scalePct?: string;
    };
}

export interface MetricsWidget extends Widget {
    content: {
        title: string;
        datasetId: string;
        oneMetricPerRow: boolean;
        datasetType?: DatasetType;
        significantDigitLabels: boolean;
        metricsCenterAlign: boolean;
        s3Key: {
            raw: string;
            json: string;
        };
    };
}

export interface SectionWidget extends Widget {
    content: {
        title: string;
        summary: string;
        widgetIds?: Array<string>;
        showWithTabs: boolean;
        horizontally?: boolean;
    };
}

export enum DatasetSchema {
    None = "None",
    Metrics = "Metrics",
}

export type Dataset = {
    id: string;
    fileName: string;
    s3Key: {
        raw: string;
        json: string;
    };
    createdBy?: string;
    updatedAt?: Date;
    sourceType?: SourceType;
    schema?: DatasetSchema;
};

export type PublicHomepage = {
    title: string;
    description: string;
    dashboards: Array<PublicDashboard>;
};

export type Homepage = {
    title: string;
    description: string;
    updatedAt: Date;
};

export type PublicSettings = {
    dateTimeFormat: {
        date: string;
        time: string;
    };
    navbarTitle?: string;
    colors: {
        primary: string;
        secondary: string;
    };
    topicAreaLabels?: {
        singular: string;
        plural: string;
    };
    customLogoS3Key?: string;
    customFaviconS3Key?: string;
    contactEmailAddress?: string;
    contactUsContent?: string;
    analyticsTrackingId?: string;
};

export type Settings = {
    publishingGuidance: string;
    dateTimeFormat: {
        date: string;
        time: string;
    };
    updatedAt?: Date;
    navbarTitle?: string;
    colors: {
        primary: string;
        secondary: string;
    };
    topicAreaLabels: {
        singular: string;
        plural: string;
    };
    customLogoS3Key?: string;
    customLogoAltText?: string;
    customFaviconS3Key?: string;
    contactEmailAddress?: string;
    adminContactEmailAddress?: string;
    contactUsContent?: string;
    analyticsTrackingId?: string;
};

export type Metric = {
    title: string;
    value: number;
    changeOverTime?: string;
    percentage?: string;
    currency?: string;
    startDate?: string;
    endDate?: string;
};

// Type for the History object in react-router. Defines the
// location state that is common across all routes.
// Usage in a component: useHistory<LocationState>();
export type LocationState = {
    alert?: {
        type: "success" | "warning" | "info" | "error";
        message: string;
        to?: string;
        linkLabel?: string;
    };
    emails?: string;
    usernames?: Array<string>;
    id?: string;
    metrics?: Array<Metric>;
    metric?: Metric;
    position?: number;
    showTitle?: boolean;
    oneMetricPerRow?: boolean;
    significantDigitLabels?: boolean;
    metricsCenterAlign?: boolean;
    metricTitle?: string;
    origin?: string;
    json?: Array<any>;
    staticDataset?: Dataset;
    redirectUrl?: string;
    crumbLabel?: string;
    datasetType?: DatasetType;
};

export enum UserRoles {
    Admin = "Admin",
    Editor = "Editor",
    /**
     * @type {string}
     * @deprecated no longer supported
     */
    Publisher = "Publisher",
    Public = "Public",
}

export type User = {
    userId: string;
    enabled: boolean;
    userStatus: string;
    sub: string;
    email: string;
    roles: Array<UserRoles>;
    createdAt: Date;
    updatedAt: Date;
};

export enum ColumnDataType {
    Number = "Number",
    Text = "Text",
    Date = "Date",
}

export enum NumberDataType {
    Percentage = "Percentage",
    Currency = "Currency",
    "With thousands separators" = "With thousands separators",
}

export enum CurrencyDataType {
    "Dollar $" = "Dollar $",
    "Euro €" = "Euro €",
    "Pound £" = "Pound £",
}

export interface ColumnMetadata {
    columnName: string;
    dataType?: ColumnDataType;
    numberType?: NumberDataType;
    currencyType?: CurrencyDataType;
    hidden: boolean;
}
