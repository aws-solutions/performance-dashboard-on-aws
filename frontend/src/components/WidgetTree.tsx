/*
 *  Copyright Amazon.com, Inc. or its affiliates. All Rights Reserved.
 *  SPDX-License-Identifier: Apache-2.0
 */

import React, { useState, useEffect } from "react";
import {
    DragDropContext,
    DragStart,
    DragUpdate,
    Droppable,
    DropResult,
    ResponderProvided,
} from "react-beautiful-dnd";
import { Widget, WidgetType } from "../models";
import WidgetTreeItem from "./WidgetTreeItem";
import OrderingService, { WidgetTreeData } from "../services/OrderingService";
import { useTranslation } from "react-i18next";
import SecondaryActionBar from "./SecondaryActionBar";
import Button from "./Button";

interface Props {
    onClick: Function;
    onDelete: (widget: Widget) => void;
    onDuplicate: (widget: Widget) => void;
    onDrag: (widgets: Widget[]) => void;
    widgets: Array<Widget>;
}

function WidgetTree(props: Props) {
    const { t } = useTranslation();
    const droppableId = "widget-tree";
    const [tree, setTree] = useState<WidgetTreeData>();
    const [isDropDisabled, setIsDropDisabled] = useState(false);

    const onDragStarted = (start: DragStart, provided: ResponderProvided) => {
        if (!tree) {
            return;
        }
        const source = tree.map[start.source.index];
        provided.announce(
            t("WidgetTree.ItemLifted", {
                label: source?.label,
                name: source?.widget?.name,
            }),
        );
    };

    const onDragUpdate = (update: DragUpdate, provided: ResponderProvided) => {
        if (!tree) {
            return;
        }

        const source = tree.map[update.source.index];
        if (
            update.source.droppableId === droppableId &&
            update.destination &&
            update.destination.droppableId === droppableId
        ) {
            const destination = tree.map[update.destination.index];
            provided.announce(
                t("WidgetTree.ItemMoved", {
                    sourceLabel: source?.label,
                    sourceName: source?.widget?.name,
                    destinationLabel: destination?.label,
                    destinationName: destination?.widget?.name,
                }),
            );
            const isDisabled =
                source &&
                destination &&
                !!destination.section &&
                source.widget?.widgetType === WidgetType.Section &&
                destination.section !== source.id;
            if (isDisabled !== isDropDisabled) {
                setIsDropDisabled(isDisabled);
            }
        } else {
            provided.announce(
                t("WidgetTree.OutsideTheDroppableAreaError", {
                    label: source?.label,
                    name: source?.widget?.name,
                }),
            );
            setIsDropDisabled(false);
        }
    };

    const moveWidget = (sourceIndex: number, destinationIndex: number): void => {
        if (!tree) {
            return;
        }
        if (sourceIndex < 0) {
            sourceIndex = 0;
        }
        if (destinationIndex >= tree.nodes.length) {
            destinationIndex = tree.nodes.length - 1;
        }

        const widgets = OrderingService.moveWidget(tree, sourceIndex, destinationIndex);
        if (widgets) {
            setTree(OrderingService.buildTree(widgets));
            props.onDrag(widgets);
        }
    };

    const onDragEnd = (result: DropResult, provided: ResponderProvided) => {
        setIsDropDisabled(false);
        const { destination, source } = result;
        if (!destination) {
            return;
        }
        if (destination.droppableId === source.droppableId && destination.index === source.index) {
            return;
        }
        const sourceItem = tree?.map[source.index];
        const destinationItem = tree?.map[destination.index];
        moveWidget(source.index, destination.index);

        provided.announce(
            t("WidgetTree.ItemDropped", {
                sourceLabel: sourceItem?.label,
                sourceName: sourceItem?.widget?.name,
                destinationLabel: destinationItem?.label,
                destinationName: destinationItem?.widget?.name,
            }),
        );
    };

    useEffect(() => {
        const newNodes = OrderingService.buildTree(props.widgets);
        setTree(newNodes);
    }, [props.widgets]);

    return (
        <DragDropContext
            onDragStart={onDragStarted}
            onDragUpdate={onDragUpdate}
            onDragEnd={onDragEnd}
        >
            {
                <Droppable droppableId={droppableId} isDropDisabled={false}>
                    {(provided) => (
                        <div
                            ref={provided.innerRef}
                            {...provided.droppableProps}
                            className="padding-top-2 padding-bottom-2"
                        >
                            <hr className="margin-top-2 border-base-lightest" />
                            <h2 className="margin-bottom-2 margin-top-2">{t("ContentItems")}</h2>
                            <ol className="widget-tree" aria-label={t("ContentItems")}>
                                {tree?.nodes.map((node) => {
                                    return (
                                        <WidgetTreeItem
                                            key={node.id}
                                            node={node}
                                            isDropDisabled={isDropDisabled}
                                            onDuplicate={props.onDuplicate}
                                            onDelete={props.onDelete}
                                            onMove={moveWidget}
                                            canMoveUp={node.dragIndex > 0}
                                            canMoveDown={node.dragIndex < tree.length - 1}
                                        />
                                    );
                                })}
                            </ol>
                            {provided.placeholder}
                            {!props.widgets || props.widgets.length === 0 ? (
                                <SecondaryActionBar className="text-center padding-5 margin-y-2">
                                    <div>
                                        <p>
                                            {t("NoContentItems")} <br />
                                            {t("ChartsTablesMore")}
                                        </p>
                                        <div className="text-center margin-top-4">
                                            <Button
                                                className="margin-top-1"
                                                variant="base"
                                                onClick={() => {
                                                    if (props.onClick) {
                                                        props.onClick();
                                                    }
                                                }}
                                                ariaLabel={t("PlusAddContentItem")}
                                            >
                                                {t("PlusAddContentItem")}
                                            </Button>
                                        </div>
                                    </div>
                                </SecondaryActionBar>
                            ) : (
                                <div className="text-center margin-top-2">
                                    <button
                                        className="usa-button usa-button--base margin-top-1"
                                        onClick={() => {
                                            if (props.onClick) {
                                                props.onClick();
                                            }
                                        }}
                                        aria-label={t("PlusAddContentItem")}
                                    >
                                        {t("PlusAddContentItem")}
                                    </button>
                                </div>
                            )}
                        </div>
                    )}
                </Droppable>
            }
        </DragDropContext>
    );
}

export default WidgetTree;
