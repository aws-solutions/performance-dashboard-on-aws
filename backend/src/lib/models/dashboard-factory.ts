import { v4 as uuidv4 } from "uuid";
import { Dashboard, DashboardItem } from "./dashboard-models";
import { User } from "./user-models";
import topicareaFactory from "./topicarea-factory";

const DASHBOARD: string = 'Dashboard';

function createNew(name: string, topicAreaId: string, topicAreaName: string, user: User): Dashboard {
  return {
    id: uuidv4(),
    name,
    topicAreaId,
    topicAreaName,
    createdBy: user.userId,
  };
}

/**
 * Converts a Dashboard to a DynamoDB item.
 */
function toItem(dashboard: Dashboard): DashboardItem {
  let item: DashboardItem = {
    pk: itemId(dashboard.id),
    sk: itemId(dashboard.id),
    type: DASHBOARD,
    dashboardName: dashboard.name,
    topicAreaName: dashboard.topicAreaName,
    topicAreaId: topicareaFactory.itemId(dashboard.topicAreaId),
    description: dashboard.description,
    overview: dashboard.overview,
    createdBy: dashboard.createdBy,
  };
  return item;
}

/**
 * Converts a DynamoDB item into a Dashboard object
 */
function fromItem(item: DashboardItem): Dashboard {
  const id = item.pk.substring(10);
  let dashboard: Dashboard = {
    id,
    name: item.dashboardName,
    topicAreaId: item.topicAreaId.substring(10),
    topicAreaName: item.topicAreaName,
    description: item.description,
    overview: item.overview,
    createdBy: item.createdBy,
  }
  return dashboard;
}

function itemId(id: string): string { return `${DASHBOARD}#${id}` }

export default {
  toItem,
  createNew,
  fromItem,
  itemId,
};
