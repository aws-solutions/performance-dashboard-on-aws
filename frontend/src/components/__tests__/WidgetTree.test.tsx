/*
 *  Copyright Amazon.com, Inc. or its affiliates. All Rights Reserved.
 *  SPDX-License-Identifier: Apache-2.0
 */

import React from "react";
import { render, fireEvent, act } from "@testing-library/react";
import { Widget } from "../../models";
import WidgetTree from "../WidgetTree";
import { MemoryRouter } from "react-router-dom";

const widgets: Array<Widget> = [
    {
        dashboardId: "abc",
        id: "123",
        name: "The benefits of bananas",
        widgetType: "Text",
        showTitle: true,
        order: 1,
        updatedAt: new Date("2020-09-22T20:13:08Z"),
        content: {},
    },
    {
        dashboardId: "abc",
        id: "456",
        name: "The benefits of wine",
        widgetType: "Text",
        showTitle: true,
        order: 2,
        updatedAt: new Date("2020-09-22T20:13:08Z"),
        content: {},
    },
];

const renderWidgetTree = (
    widgets: Array<Widget>,
    {
        onClick,
        onDuplicate,
        onDelete,
        onDrag,
    }: {
        onClick?: Function;
        onDuplicate?: Function;
        onDelete?: Function;
        onDrag?: Function;
    },
) => {
    return render(
        <WidgetTree
            widgets={widgets}
            onClick={onClick ?? jest.fn()}
            onDuplicate={onDuplicate ?? jest.fn()}
            onDelete={onDelete ?? jest.fn()}
            onDrag={onDrag ?? jest.fn()}
        />,
        {
            wrapper: MemoryRouter,
        },
    );
};

test("renders an empty box", async () => {
    const onClick = jest.fn();
    const wrapper = renderWidgetTree(widgets, { onClick });
    expect(wrapper.container).toMatchSnapshot();
});

test("calls the onClick callback when button is pressed", async () => {
    const onClick = jest.fn();
    const wrapper = renderWidgetTree(widgets, { onClick });
    fireEvent.click(wrapper.getByText("+ Add content item"));
    expect(onClick).toBeCalled();
});

test("calls onDelete when user clicks delete button", async () => {
    const onDelete = jest.fn().mockImplementation(() => {
        console.log("!!!!!!!!!!!!!!!!!DELETED!!!!!!!!!!!!!!!!!!!");
    });
    const onClick = jest.fn();
    const wrapper = renderWidgetTree(widgets, { onClick, onDelete });

    await act(async () => {
        fireEvent.click(wrapper.getByLabelText("Actions for The benefits of bananas"));
    });

    await act(async () => {
        fireEvent.click(wrapper.getByLabelText("Delete The benefits of bananas"));
    });

    expect(onDelete).toBeCalled();
});
