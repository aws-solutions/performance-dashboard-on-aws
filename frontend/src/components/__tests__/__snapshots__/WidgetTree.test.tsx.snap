// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`renders an empty box 1`] = `
<div>
  <div
    class="padding-top-2 padding-bottom-2"
    data-rbd-droppable-context-id="0"
    data-rbd-droppable-id="widget-tree"
  >
    <hr
      class="margin-top-2 border-base-lightest"
    />
    <h2
      class="margin-bottom-2 margin-top-2"
    >
      Content Items
    </h2>
    <ol
      aria-label="Content Items"
      class="widget-tree"
    >
      <li
        aria-label="The benefits of bananas"
        class="nostyle"
      >
        <div
          class=""
        >
          <div
            class="margin-top-1 "
            data-rbd-draggable-context-id="0"
            data-rbd-draggable-id="123"
          >
            <div
              class="border-base-lighter border-1px shadow-1 z-200 radius-lg bg-white grid-col "
            >
              <div
                class="grid-row flex-1"
              >
                <div
                  class="grid-row grid-col flex-1 padding-1"
                >
                  <div
                    aria-describedby="rbd-hidden-text-0-hidden-text-0"
                    class="text-base-darker grid-col flex-3 text-center display-flex flex-align-center flex-justify-center margin-left-2 margin-right-1 margin-top-1 margin-bottom-1"
                    data-rbd-drag-handle-context-id="0"
                    data-rbd-drag-handle-draggable-id="123"
                    draggable="false"
                    role="button"
                    tabindex="0"
                  >
                    <svg
                      aria-hidden="true"
                      class="svg-inline--fa fa-grip-vertical fa-w-10 fa-1x "
                      data-icon="grip-vertical"
                      data-prefix="fas"
                      focusable="false"
                      role="img"
                      viewBox="0 0 320 512"
                      xmlns="http://www.w3.org/2000/svg"
                    >
                      <path
                        d="M96 32H32C14.33 32 0 46.33 0 64v64c0 17.67 14.33 32 32 32h64c17.67 0 32-14.33 32-32V64c0-17.67-14.33-32-32-32zm0 160H32c-17.67 0-32 14.33-32 32v64c0 17.67 14.33 32 32 32h64c17.67 0 32-14.33 32-32v-64c0-17.67-14.33-32-32-32zm0 160H32c-17.67 0-32 14.33-32 32v64c0 17.67 14.33 32 32 32h64c17.67 0 32-14.33 32-32v-64c0-17.67-14.33-32-32-32zM288 32h-64c-17.67 0-32 14.33-32 32v64c0 17.67 14.33 32 32 32h64c17.67 0 32-14.33 32-32V64c0-17.67-14.33-32-32-32zm0 160h-64c-17.67 0-32 14.33-32 32v64c0 17.67 14.33 32 32 32h64c17.67 0 32-14.33 32-32v-64c0-17.67-14.33-32-32-32zm0 160h-64c-17.67 0-32 14.33-32 32v64c0 17.67 14.33 32 32 32h64c17.67 0 32-14.33 32-32v-64c0-17.67-14.33-32-32-32z"
                        fill="currentColor"
                      />
                    </svg>
                  </div>
                  <div
                    class="grid-col flex-6 text-center display-flex flex-align-center flex-justify-center font-sans-md margin-left-2 margin-top-1 margin-bottom-1"
                  >
                    1
                  </div>
                  <div
                    class="grid-col flex-4 grid-row flex-column text-center margin-left-2 margin-right-2"
                  >
                    <div
                      class="grid-col flex-6"
                    >
                      <br />
                    </div>
                    <div
                      class="grid-col flex-6"
                    >
                      <button
                        aria-label="Move The benefits of bananas down"
                        class="usa-button usa-button--unstyled text-base-darker hover:text-base-darkest active:text-base-darkest"
                      >
                        <svg
                          aria-hidden="true"
                          class="svg-inline--fa fa-arrow-down fa-w-14 fa-xs "
                          data-icon="arrow-down"
                          data-prefix="fas"
                          focusable="false"
                          id="123-move-down"
                          role="img"
                          viewBox="0 0 448 512"
                          xmlns="http://www.w3.org/2000/svg"
                        >
                          <path
                            d="M413.1 222.5l22.2 22.2c9.4 9.4 9.4 24.6 0 33.9L241 473c-9.4 9.4-24.6 9.4-33.9 0L12.7 278.6c-9.4-9.4-9.4-24.6 0-33.9l22.2-22.2c9.5-9.5 25-9.3 34.3.4L184 343.4V56c0-13.3 10.7-24 24-24h32c13.3 0 24 10.7 24 24v287.4l114.8-120.5c9.3-9.8 24.8-10 34.3-.4z"
                            fill="currentColor"
                          />
                        </svg>
                      </button>
                    </div>
                  </div>
                </div>
                <div
                  class="border-base-lighter border-left"
                />
                <div
                  class="grid-col flex-11 grid-row padding-1 margin-y-1"
                >
                  <div
                    class="grid-col flex-8 usa-tooltip"
                    data-position="bottom"
                    title="The benefits of bananas"
                  >
                    <div
                      class="margin-left-1 text-wrap text-bold"
                    >
                      <a
                        aria-label="Edit The benefits of bananas"
                        class="usa-link action-link"
                        href="/admin/dashboard/abc/edit-text/123"
                        target=""
                      >
                        <span>
                          The benefits of bananas
                        </span>
                      </a>
                    </div>
                  </div>
                  <div
                    class="grid-col flex-3 text-italic text-right"
                  >
                    (Text)
                  </div>
                  <div
                    class="grid-col flex-1 margin-left-2 margin-right-2 text-right"
                  >
                    <button
                      aria-controls="menu--1"
                      aria-haspopup="true"
                      aria-label="Actions for The benefits of bananas"
                      class="usa-button usa-button--unstyled"
                      data-reach-menu-button=""
                      id="menu-button--menu--1"
                      type="button"
                    >
                      
                      <svg
                        aria-hidden="true"
                        class="svg-inline--fa fa-ellipsis-v fa-w-6 margin-left-1"
                        data-icon="ellipsis-v"
                        data-prefix="fas"
                        focusable="false"
                        role="img"
                        viewBox="0 0 192 512"
                        xmlns="http://www.w3.org/2000/svg"
                      >
                        <path
                          d="M96 184c39.8 0 72 32.2 72 72s-32.2 72-72 72-72-32.2-72-72 32.2-72 72-72zM24 80c0 39.8 32.2 72 72 72s72-32.2 72-72S135.8 8 96 8 24 40.2 24 80zm0 352c0 39.8 32.2 72 72 72s72-32.2 72-72-32.2-72-72-72-72 32.2-72 72z"
                          fill="currentColor"
                        />
                      </svg>
                    </button>
                  </div>
                </div>
              </div>
            </div>
          </div>
        </div>
      </li>
      <li
        aria-label="The benefits of wine"
        class="nostyle"
      >
        <div
          class=""
        >
          <div
            class="margin-top-1 "
            data-rbd-draggable-context-id="0"
            data-rbd-draggable-id="456"
          >
            <div
              class="border-base-lighter border-1px shadow-1 z-200 radius-lg bg-white grid-col "
            >
              <div
                class="grid-row flex-1"
              >
                <div
                  class="grid-row grid-col flex-1 padding-1"
                >
                  <div
                    aria-describedby="rbd-hidden-text-0-hidden-text-0"
                    class="text-base-darker grid-col flex-3 text-center display-flex flex-align-center flex-justify-center margin-left-2 margin-right-1 margin-top-1 margin-bottom-1"
                    data-rbd-drag-handle-context-id="0"
                    data-rbd-drag-handle-draggable-id="456"
                    draggable="false"
                    role="button"
                    tabindex="0"
                  >
                    <svg
                      aria-hidden="true"
                      class="svg-inline--fa fa-grip-vertical fa-w-10 fa-1x "
                      data-icon="grip-vertical"
                      data-prefix="fas"
                      focusable="false"
                      role="img"
                      viewBox="0 0 320 512"
                      xmlns="http://www.w3.org/2000/svg"
                    >
                      <path
                        d="M96 32H32C14.33 32 0 46.33 0 64v64c0 17.67 14.33 32 32 32h64c17.67 0 32-14.33 32-32V64c0-17.67-14.33-32-32-32zm0 160H32c-17.67 0-32 14.33-32 32v64c0 17.67 14.33 32 32 32h64c17.67 0 32-14.33 32-32v-64c0-17.67-14.33-32-32-32zm0 160H32c-17.67 0-32 14.33-32 32v64c0 17.67 14.33 32 32 32h64c17.67 0 32-14.33 32-32v-64c0-17.67-14.33-32-32-32zM288 32h-64c-17.67 0-32 14.33-32 32v64c0 17.67 14.33 32 32 32h64c17.67 0 32-14.33 32-32V64c0-17.67-14.33-32-32-32zm0 160h-64c-17.67 0-32 14.33-32 32v64c0 17.67 14.33 32 32 32h64c17.67 0 32-14.33 32-32v-64c0-17.67-14.33-32-32-32zm0 160h-64c-17.67 0-32 14.33-32 32v64c0 17.67 14.33 32 32 32h64c17.67 0 32-14.33 32-32v-64c0-17.67-14.33-32-32-32z"
                        fill="currentColor"
                      />
                    </svg>
                  </div>
                  <div
                    class="grid-col flex-6 text-center display-flex flex-align-center flex-justify-center font-sans-md margin-left-2 margin-top-1 margin-bottom-1"
                  >
                    2
                  </div>
                  <div
                    class="grid-col flex-4 grid-row flex-column text-center margin-left-2 margin-right-2"
                  >
                    <div
                      class="grid-col flex-6"
                    >
                      <button
                        aria-label="Move The benefits of wine up"
                        class="usa-button usa-button--unstyled text-base-darker hover:text-base-darkest active:text-base-darkest"
                      >
                        <svg
                          aria-hidden="true"
                          class="svg-inline--fa fa-arrow-up fa-w-14 fa-xs "
                          data-icon="arrow-up"
                          data-prefix="fas"
                          focusable="false"
                          id="456-move-up"
                          role="img"
                          viewBox="0 0 448 512"
                          xmlns="http://www.w3.org/2000/svg"
                        >
                          <path
                            d="M34.9 289.5l-22.2-22.2c-9.4-9.4-9.4-24.6 0-33.9L207 39c9.4-9.4 24.6-9.4 33.9 0l194.3 194.3c9.4 9.4 9.4 24.6 0 33.9L413 289.4c-9.5 9.5-25 9.3-34.3-.4L264 168.6V456c0 13.3-10.7 24-24 24h-32c-13.3 0-24-10.7-24-24V168.6L69.2 289.1c-9.3 9.8-24.8 10-34.3.4z"
                            fill="currentColor"
                          />
                        </svg>
                      </button>
                    </div>
                    <div
                      class="grid-col flex-6"
                    >
                      <br />
                    </div>
                  </div>
                </div>
                <div
                  class="border-base-lighter border-left"
                />
                <div
                  class="grid-col flex-11 grid-row padding-1 margin-y-1"
                >
                  <div
                    class="grid-col flex-8 usa-tooltip"
                    data-position="bottom"
                    title="The benefits of wine"
                  >
                    <div
                      class="margin-left-1 text-wrap text-bold"
                    >
                      <a
                        aria-label="Edit The benefits of wine"
                        class="usa-link action-link"
                        href="/admin/dashboard/abc/edit-text/456"
                        target=""
                      >
                        <span>
                          The benefits of wine
                        </span>
                      </a>
                    </div>
                  </div>
                  <div
                    class="grid-col flex-3 text-italic text-right"
                  >
                    (Text)
                  </div>
                  <div
                    class="grid-col flex-1 margin-left-2 margin-right-2 text-right"
                  >
                    <button
                      aria-controls="menu--2"
                      aria-haspopup="true"
                      aria-label="Actions for The benefits of wine"
                      class="usa-button usa-button--unstyled"
                      data-reach-menu-button=""
                      id="menu-button--menu--2"
                      type="button"
                    >
                      
                      <svg
                        aria-hidden="true"
                        class="svg-inline--fa fa-ellipsis-v fa-w-6 margin-left-1"
                        data-icon="ellipsis-v"
                        data-prefix="fas"
                        focusable="false"
                        role="img"
                        viewBox="0 0 192 512"
                        xmlns="http://www.w3.org/2000/svg"
                      >
                        <path
                          d="M96 184c39.8 0 72 32.2 72 72s-32.2 72-72 72-72-32.2-72-72 32.2-72 72-72zM24 80c0 39.8 32.2 72 72 72s72-32.2 72-72S135.8 8 96 8 24 40.2 24 80zm0 352c0 39.8 32.2 72 72 72s72-32.2 72-72-32.2-72-72-72-72 32.2-72 72z"
                          fill="currentColor"
                        />
                      </svg>
                    </button>
                  </div>
                </div>
              </div>
            </div>
          </div>
        </div>
      </li>
    </ol>
    <div
      class="text-center margin-top-2"
    >
      <button
        aria-label="+ Add content item"
        class="usa-button usa-button--base margin-top-1"
      >
        + Add content item
      </button>
    </div>
  </div>
</div>
`;
